File: Final/Code/Project_4/BaiToan7/top_down/top_down.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct TreeNode {
    string label;
    vector<TreeNode*> children;
    TreeNode(string val) : label(val) {}
};

void top_down(TreeNode* node) {
    if (!node) return;
    queue<TreeNode*> q;
    q.push(node);
    while (!q.empty()) {
        TreeNode* curr = q.front(); q.pop();
        cout << curr->label << " ";
        for (TreeNode* child : curr->children)
            q.push(child);
    }
}

int main() {
    TreeNode* root = new TreeNode("A");
    TreeNode* B = new TreeNode("B");
    TreeNode* C = new TreeNode("C");
    TreeNode* D = new TreeNode("D");
    TreeNode* E = new TreeNode("E");
    TreeNode* F = new TreeNode("F");
    TreeNode* G = new TreeNode("G");
    TreeNode* H = new TreeNode("H");

    root->children = {B, C, D};
    B->children = {E, F};
    C->children = {H};
    D->children = {G};

    top_down(root);
}

File: Final/Code/Project_4/BaiToan7/top_down/top_down.py
from collections import deque

class TreeNode:
    def __init__(self, label):
        self.label = label
        self.children = []

def top_down(node):
    if node is None:
        return
    queue = deque([node])
    while queue:
        curr = queue.popleft()
        print(curr.label, end=' ')
        for child in curr.children:
            queue.append(child)

root = TreeNode("A")
root.children = [TreeNode("B"), TreeNode("C"), TreeNode("D")]
root.children[0].children = [TreeNode("E"), TreeNode("F")]
root.children[1].children = [TreeNode("H")]
root.children[2].children = [TreeNode("G")]

top_down(root)